var/spack/repos/builtin/packages/flecsi/package.py

```diff
@@ -130,13 +135,13 @@ def cmake_args(self):
                 self.define_from_variant("ENABLE_DOCUMENTATION", "doc"),
             ]
 
-            if self.spec.satisfies("+rocm"):
+            if self.spec.satisfies("^kokkos +rocm"):
                 options.append(self.define("CMAKE_CXX_COMPILER", self.spec["hip"].hipcc))
                 options.append(self.define("CMAKE_C_COMPILER", self.spec["hip"].hipcc))
                 if self.spec.satisfies("backend=legion"):
                     # CMake pulled in via find_package(Legion) won't work without this
                     options.append(self.define("HIP_PATH", "{0}/hip".format(spec["hip"].prefix)))
-            elif self.spec.satisfies("+kokkos"):
+            elif self.spec.satisfies("^kokkos"):
```

Comment:
For the old versions that can have `~kokkos`, wouldn't this trigger incorrectly for `sim ^flecsi ~kokkos ^kokkos`?  Maybe we don't care about that case, though.

Reply:
No. `^kokkos` is only satisfied if it is in the dependency tree of `flecsi`, but not if it's only in the `sim` tree.

Reply:
Good, but can a user even specify that?  The overall spec syntax isn't hierarchical like that.

Reply:
We can generate a case where `sim` has a hard kokkos and flecsi dependency, but with the old versions if you specify `^flecsi ~kokkos` any `^kokkos` will only apply to the `sim` dependency tree.

Don't confuse spec syntax of the user with what is being checked inside of the package. `spec.satisfies` uses that syntax, but it doesn't mean it has to match anything the user explicitly specifies. We can have Spack environments with only `sim` as main spec, along with package configurations with preferences and requires that set a certain configuration. `spec.satisfies` allows you to evaluate a condition on the resulting *generated* dependency tree, with the limitation of it being the dependency (sub)tree of that particular package (`self`).

Reply:
Sure: `spec.satisfies` sees the appropriate subset; I was just concerned about possible circularity in determining the contents of that subset _from_ the user spec.  Fortunately, `~kokkos ^kokkos` (as regards FleCSI's part of the spec) is impossible in all versions (sometimes because `~kokkos` is itself meaningless), so we don't have to worry about what the dependencies mean then.

Reply:
Can we resolve this?

Reply:
Sure, but I don't have that button.